lib/krb5/init_creds_pw.c

```diff
@@ -2758,62 +3309,102 @@ init_creds_step(krb5_context context,
 		if (ret)
 		    goto out;
 
+		gettimeofday(&start_time, NULL);
+
 		krb5_init_creds_set_password(context, ctx, buf2);
 
- 		ctx->used_pa_types = 0;
-		ret = 0;
+		pa_restart(context, ctx);
+
+	    } else if (ret == KRB5KDC_ERR_PREAUTH_FAILED) {
+
+		/*
+		 * Old MIT KDC can't handle KRB5_PADATA_REQ_ENC_PA_REP,
+		 * so drop it and try again. But only try that for MIT
+		 * Kerberos servers by keying of no METHOD-DATA.
+		 */
+		if (ctx->runflags.allow_enc_pa_rep) {
+		    if (ctx->md.len != 0) {
+			_krb5_debug(context, 10, "Server sent PA data with KRB-ERROR, "
+				    "so not a pre 1.7 MIT KDC and won't retry w/o ENC-PA-REQ");
+			goto out;
+		    }
+		    _krb5_debug(context, 10, "Disabling allow_enc_pa_rep and trying again");
+		    ctx->runflags.allow_enc_pa_rep = 0;
+		    goto retry;
+		}
+
+		if (ctx->fast_state.flags & KRB5_FAST_DISABLED) {
+		    _krb5_debug(context, 10, "FAST disabled and got preauth failed");
+		    goto out;
+		}
+
+		if ((ctx->fast_state.flags & KRB5_FAST_OPTIMISTIC) == 0) {
+		    _krb5_debug(context, 10, "Preauth failed");
+		    goto out;
+		}
+
+		_krb5_debug(context, 10, "preauth failed with Optimistic FAST, trying w/o FAST");
+
+		ctx->fast_state.flags &= ~KRB5_FAST_OPTIMISTIC;
+		ctx->fast_state.flags |= KRB5_FAST_DISABLED;
 
- 	    } else if (ret == KRB5KDC_ERR_PREAUTH_FAILED) {
- 
- 		if (ctx->fast_state.flags & KRB5_FAST_DISABLED)
- 		    goto out;
-		if (!(ctx->fast_state.flags & KRB5_FAST_EXPECTED)) {
+	    retry:
+		pa_restart(context, ctx);
+
+	    } else {
+		if (ctx->fast_state.flags & KRB5_FAST_OPTIMISTIC) {
+		    _krb5_debug(context, 10,
+				"Some other error %d failed with Optimistic FAST, trying w/o FAST", ret);
+
+		    ctx->fast_state.flags &= ~KRB5_FAST_OPTIMISTIC;
+		    ctx->fast_state.flags |= KRB5_FAST_DISABLED;
+		    pa_restart(context, ctx);
+
```

Comment:
Should there be an `} else {` here? I noticed this was causing this function to keep looping when it received an error from the KDC that wasn't handled in one of the cases above.

Reply:
Yes, how about this:

```diff
diff --git a/lib/krb5/init_creds_pw.c b/lib/krb5/init_creds_pw.c
index de6b13a53..b04b6ebab 100644
--- a/lib/krb5/init_creds_pw.c
+++ b/lib/krb5/init_creds_pw.c
@@ -3351,18 +3351,16 @@ init_creds_step(krb5_context context,
            retry:
                pa_restart(context, ctx);
 
-           } else {
-               if (ctx->fast_state.flags & KRB5_FAST_OPTIMISTIC) {
-                   _krb5_debug(context, 10,
-                               "Some other error %d failed with Optimistic FAST, trying w/o FAST", ret);
-
-                   ctx->fast_state.flags &= ~KRB5_FAST_OPTIMISTIC;
-                   ctx->fast_state.flags |= KRB5_FAST_DISABLED;
-                   pa_restart(context, ctx);
+           } else if (ctx->fast_state.flags & KRB5_FAST_OPTIMISTIC) {
+               _krb5_debug(context, 10,
+                           "Some other error %d failed with Optimistic FAST, trying w/o FAST", ret);
 
-                   /* some other error code from the KDC, lets return it to the user */
-                   goto out;
-               }
+               ctx->fast_state.flags &= ~KRB5_FAST_OPTIMISTIC;
+               ctx->fast_state.flags |= KRB5_FAST_DISABLED;
+               pa_restart(context, ctx);
+           } else {
+               /* some other error code from the KDC, lets return it to the user */
+               goto out;
            }
        }
     }
```